tensorflow/python/ops/array_ops.py

```diff
@@ -1907,18 +1907,35 @@ def one_hot(indices, depth, on_value=1, off_value=0,
     output: The one-hot tensor.
 
   Raises:
-    TypeError: If dtype is `tf.string`
+    TypeError: If dtype of provided `on_value` and `off_value` don't match
   """
-  # Check for bad dtype specification
-  if dtype == dtypes.string:
-    raise TypeError("dtype must be a numeric type")
-
```

Comment:
first, check if dtype is None [and on_value and off_value are None].  if so, set dtype to float32.

next you can check "if on_value is None" and set it to 1.0 via convert_to_tensor; same with off_value.

then check to make sure their dtypes match.